edx_proctoring/api.py

```diff
@@ -0,0 +1,777 @@
+    existing_attempt = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+    if existing_attempt:
+        if existing_attempt.is_sample_attempt:
+            # Archive the existing attempt by deleting it.
+            existing_attempt.delete_exam_attempt()
+        else:
+            err_msg = (
+                'Cannot create new exam attempt for exam_id = {exam_id} and '
+                'user_id = {user_id} because it already exists!'
+            ).format(exam_id=exam_id, user_id=user_id)
+
+            raise StudentExamAttemptAlreadyExistsException(err_msg)
+
+    allowed_time_limit_mins = exam['time_limit_mins']
+
+    # add in the allowed additional time
+    allowance = ProctoredExamStudentAllowance.get_allowance_for_user(
+        exam_id,
+        user_id,
+        "Additional time (minutes)"
+    )
+
+    if allowance:
+        allowance_extra_mins = int(allowance.value)
+        allowed_time_limit_mins += allowance_extra_mins
+
+    attempt_code = unicode(uuid.uuid4()).upper()
+
+    external_id = None
+    if taking_as_proctored:
+        callback_url = 'http://{hostname}{path}'.format(
+            hostname=settings.SITE_NAME,
+            path=reverse(
+                'edx_proctoring.anonymous.proctoring_launch_callback.start_exam',
+                args=[attempt_code]
+            )
+        )
+
+        # get the name of the user, if the service is available
+        full_name = None
+
+        credit_service = get_runtime_service('credit')
+        if credit_service:
+            credit_state = credit_service.get_credit_state(user_id, exam['course_id'])
+            full_name = credit_state['profile_fullname']
+
+        # now call into the backend provider to register exam attempt
+        external_id = get_backend_provider().register_exam_attempt(
+            exam,
+            context={
+                'time_limit_mins': allowed_time_limit_mins,
+                'attempt_code': attempt_code,
+                'is_sample_attempt': exam['is_practice_exam'],
+                'callback_url': callback_url,
+                'full_name': full_name,
+            }
+        )
+
+    attempt = ProctoredExamStudentAttempt.create_exam_attempt(
+        exam_id,
+        user_id,
+        '',  # student name is TBD
+        allowed_time_limit_mins,
+        attempt_code,
+        taking_as_proctored,
+        exam['is_practice_exam'],
+        external_id
+    )
+    return attempt.id
+
+
+def start_exam_attempt(exam_id, user_id):
+    """
+    Signals the beginning of an exam attempt for a given
+    exam_id. If one already exists, then an exception should be thrown.
+
+    Returns: exam_attempt_id (PK)
+    """
+
+    existing_attempt = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+
+    if not existing_attempt:
+        err_msg = (
+            'Cannot start exam attempt for exam_id = {exam_id} '
+            'and user_id = {user_id} because it does not exist!'
```

Comment:
no, these are logging message (for debugging). I'm not sure how to give you a guidance on which are user facing, but everything user facing are in *.html or *.underscore or *.js (minimally) files. You can pass over *.py files.